From aric@sdgsystems.com  Fri Jul 31 09:08:46 2009
From: "Aric D. Blumer" <aric@sdgsystems.com>
Date: Thu, 30 Jul 2009 13:42:30 -0400
Subject: USB: ohci-pxa27x: Reconfigure power settings on resume
To: linux-usb@vger.kernel.org
Message-ID: <4A71DB86.8050701@sdgsystems.com>


From: Aric D. Blumer <aric@sdgsystems.com>

On resume, the power-related bits in UHCRHDA were not being set, so
they would default to the reset state.  For PXA3xx devices, OCPM must
be cleared, but it was remaining set from resume reset.

Signed-off-by: Aric D. Blumer <aric@sdgsystems.com>
Cc: David Brownell <dbrownell@users.sourceforge.net>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>


---
 drivers/usb/host/ohci-pxa27x.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

--- a/drivers/usb/host/ohci-pxa27x.c
+++ b/drivers/usb/host/ohci-pxa27x.c
@@ -502,6 +502,9 @@ static int ohci_hcd_pxa27x_drv_resume(st
 	struct usb_hcd *hcd = platform_get_drvdata(pdev);
 	struct pxa27x_ohci *ohci = to_pxa27x_ohci(hcd);
 	int status;
+	struct pxaohci_platform_data *inf;
+
+	inf = pdev->dev.platform_data;
 
 	if (time_before(jiffies, ohci->ohci.next_statechange))
 		msleep(5);
@@ -510,6 +513,13 @@ static int ohci_hcd_pxa27x_drv_resume(st
 	if ((status = pxa27x_start_hc(ohci, &pdev->dev)) < 0)
 		return status;
 
+	if (inf) {
+		/* Select Power Management Mode */
+		pxa27x_ohci_select_pmm(ohci, inf->port_mode);
+	} else {
+		pr_err("No platform info on resume");
+	}
+
 	ohci_finish_controller_resume(hcd);
 	return 0;
 }
